pkg/scheduler/dependency_graph.go

```diff
@@ -434,37 +482,74 @@ func (sched *Scheduler) BuildDependencyGraph(options interfaces.DependencyGraphO
 	depGraph := NewDependencyGraph(sched, options)
 	rootContext := &GraphContext{scheduler: sched, graph: depGraph, flow: flow, args: options.Args}
 
-	if _, ok := dependencies[fullFlowName]; !ok {
-		log.Printf("Flow %s is empty", options.FlowName)
-		return depGraph, nil
+	replicas, deleteReplicas, err := sched.allocateReplicas(flow, options)
+	if err != nil {
+		return nil, err
 	}
 
-	type Block struct {
-		dependency        client.Dependency
-		scheduledResource *ScheduledResource
-		parentContext     *GraphContext
+	err = sched.fillDependencyGraph(rootContext, resDefs, dependencies, flow, replicas, false)
+	if err != nil {
+		return nil, err
 	}
-
-	replicas, err := sched.allocateReplicas(flow, options)
+	err = sched.fillDependencyGraph(rootContext, resDefs, dependencies, flow, deleteReplicas, true)
 	if err != nil {
 		return nil, err
 	}
 
+	for _, value := range depGraph.graph {
+		value.RequiredBy = unique(value.RequiredBy)
+		value.Requires = unique(value.Requires)
+		value.usedInReplicas = unique(value.usedInReplicas)
+	}
+
+	if len(deleteReplicas) > 0 {
```

Comment:
So this looks like we need another abstraction there, so `BuildDependencyGraph` should be a method on struct `DependencyGraph`, where we could store mentioned above `getCorrectFlow` and other substeps as named blocks, and where mentioned lots of parameters can be passed around as struct members.

Also this would divide whole func into smaller parts, better unit testable and better understandable.